### PR TITLE
Add ability to interact with backups.

### PR DESCRIPTION
**Pull request by evanlucas**

### CODE REVIEW

**Comment by evanlucas** [comment]
Still could probably use some more tests

**Comment by evanlucas** [comment]
Related: https://github.com/rackspace/php-opencloud/issues/337

File: lib/OpenCloud/Database/Resource/Backup.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+class Backup extends PersistentResource
+{
+    public $id;
+    public $name;
+    public $description;
+    public $created;
+    public $datastore;
+    public $updated;
+    public $instance;
+    public $instance_id;
```
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Backup.php]: This needs to be camel case (`instanceId`)

File: lib/OpenCloud/Database/Resource/Backup.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    protected static $url_resource = 'backups';
+
+    protected $createKeys = array(
+        'name',
+        'instance',
+        'description'
+    );
+
+    protected $associatedResources = array(
+        'instance_id' => 'Instance'
```
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Backup.php]: Shouldn't the key be `instance`, since it represents an instance object? `instance_id` will be a string.

File: lib/OpenCloud/Database/Resource/Backup.php [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    public function __construct(Service $service, $info = null)
+    {
+        $this->instance = new \stdClass;
+        return parent::__construct($service, $info);
+    }
+
+    /**
+     * Returns the JSON object for creating the backup
+     */
+    protected function createJson()
```
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Backup.php]: Instead of asking for an object, which is harder to set up, maybe we could just ask the user for an `instanceId` string? You'd need to replace `instance` with `instanceId` in the a `createKeys` property. You can also delete this `createJson` method, since AFAIK the [base method](https://github.com/rackspace/php-opencloud/blob/working/lib/OpenCloud/Common/Resource/PersistentResource.php#L208-L230) already does what you need it to.

File: lib/OpenCloud/Database/Resource/Instance.php [hunk 4]
```diff
@@ -39,6 +39,7 @@ class Instance extends NovaResource
     public $created;
     public $updated;
     public $flavor;
+    public $backupRef;
```
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Instance.php]: The problem here is with population. Since the JSON is structured `"restorePoint": {"backupRef": "foo"}`, iterators for example won't know what to do with the value when a DB instance is instantiated in a collection. One thing we can do is extend the base [populate method](https://github.com/rackspace/php-opencloud/blob/working/lib/OpenCloud/Common/Base.php#L292) like so:

```php
public function populate($info, $setObjects = true)
{
   parent::populate($info, $setObjects);

   if (is_object($info)) { 
      $info = (array) $info;
   }

   if (isset($info['restorePoint']['backupRef'])) {
      $this->backupRef = $info['restorePoint']['backupRef'];
   }
}
```

File: lib/OpenCloud/Database/Resource/Instance.php [hunk 5]
```diff
@@ -190,13 +201,21 @@ protected function createJson()
             );
         }
 
-        return (object) array(
-            'instance' => (object) array(
+        $out = array(
```
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Instance.php]: This is just a nit pick, but since we bumped the minimum PHP version to 5.4 recently, we can use short array syntax :)

File: lib/OpenCloud/Database/Resource/Instance.php [hunk 6]
```diff
@@ -190,13 +201,21 @@ protected function createJson()
...
+        if (isset($this->backupRef)) {
+            $out['instance']['restorePoint'] = array(
+                'backupRef' => $this->backupRef
+            );
+        }
+
+        return (object) $out;
     }
 
     /**
```
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Instance.php]: It would be awesome if we had a convenience method that allowed uses to create a backup on a DB instance _itself_, rather than having to construct it. For example:

```php
$instance->createBackup();
```

you could just wrap the above method.

File: tests/OpenCloud/Tests/Database/Resource/BackupTest.php [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+use OpenCloud\Tests\Database\DatabaseTestCase;
+
+class BackupTest extends DatabaseTestCase
+{
+    public function test_Class()
+    {
+        $this->assertInstanceOf('OpenCloud\Database\Resource\Backup', $this->backup);
+    }
+
+    public function testDelete()
```
Comment (jamiehannaford) [inline tests/OpenCloud/Tests/Database/Resource/BackupTest.php]: maybe also test backup creation? specifically hitting those edge cases that will throw errors

File: tests/OpenCloud/Tests/Database/Resource/InstanceTest.php [hunk 8]
```diff
@@ -91,4 +91,9 @@ public function testUserList()
     {
         $this->assertInstanceOf(self::COLLECTION_CLASS, $this->instance->userList());
     }
+
+    public function testBackupList()
+    {
```
Comment (jamiehannaford) [inline tests/OpenCloud/Tests/Database/Resource/InstanceTest.php]: can we also add a separate test for creating a DB instance with a restore point?

**Comment by jamiehannaford** [comment]
@evanlucas :+1: This is really good, thank you so much for submitting! I've annotated a lot of stuff with some thoughts - if anything's unclear feel to comment on it and I'll elaborate on what I mean

File: lib/OpenCloud/Database/Resource/Instance.php (hunk 6 above)
Comment (evanlucas) [inline lib/OpenCloud/Database/Resource/Instance.php]: Ok, added. I did not add any checks for required params as they are checked in `Backup::create`. Is that acceptable?

File: lib/OpenCloud/Database/Resource/Backup.php (hunk 3 above)
Comment (evanlucas) [inline lib/OpenCloud/Database/Resource/Backup.php]: If I were to remove the createJson, how would we validate that the required parameters are passed?

Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Backup.php]: Sorry, you're right - I thought the base method did validation too, but it doesn't. you can probably leave your method here, add in validation, but delegate back to the base method for all the population stuff

File: lib/OpenCloud/Database/Resource/Instance.php (hunk 6 above)
Comment (jamiehannaford) [inline lib/OpenCloud/Database/Resource/Instance.php]: yep, sounds good

**Comment by evanlucas** [comment]
Strange, apparently, the API does not accept the description if it is `null`. It is listed as optional in the API docs, but if the key exists, but is `null`, the call fails

**Comment by evanlucas** [comment]
Ok, just added a fix to account for the description issue. 

All other comments should be addressed. :]

**Comment by jamiehannaford** [comment]
@evanlucas This is perfect, thanks for all your hard work! :rocket:

**Comment by evanlucas** [comment]
Glad to help out :]
